File: 31-next-permutation/31-next-permutation.rb
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def next_permutation(nums)

  i = nums.size - 2
  while i >= 0 && nums[i+1] <= nums[i]
    i -= 1
  end
  if i >= 0
    j = nums.size - 1
    while nums[j] <= nums[i]
      j -= 1
    end
    nums[i],nums[j] = nums[j],nums[i]
  end
  i += 1
  j = nums.size - 1
  while i < j
    nums[i],nums[j] = nums[j],nums[i]
    i += 1;
    j -= 1;
  end
  nums
end

File: 31-next-permutation/31-next-permutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    let i = nums.length - 2;
    while(i >= 0 && nums[i] >= nums[i + 1]){
        i--;
    }
    if(i >= 0){
        let j = nums.length - 1;
        while(j >= 0 && nums[j] <= nums[i]){
            j--;
        }
        let temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    let left = i + 1;
    let right = nums.length - 1;
    while(left < right){
        let temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
        left++;
        right--;
    }
    
};